### ISSUE TITLE
std.crypto: Aes128Gcm fails to decrypt due to AuthenticationFailed error.

### ISSUE THREAD

**Issue by naoki9911** [ziglang/zig#13506]
### Zig Version

0.11.0-dev.116+41b7e40d7

### Steps to Reproduce and Observed Behavior

Running below code with `zig test` and it causes AuthenticationFailed error.
This issue seems to be caused after [this commit(7d48cb1)](https://github.com/ziglang/zig/commit/7d48cb11389f5dfe73d25e59c1038398a8ba9ca9)
```zig
const std = @import("std");

test "Aes128Gcm" {
    const c_hex = "d1ff334a56f5bff6594a07cc87b580233f500f45e489e7f33af35edf7869fcf40aa40aa2b8ea73f848a7ca07612ef9f945cb960b4068905123ea78b111b429ba9191cd05d2a389280f526134aadc7fc78c4b729df828b5ecf7b13bd9aefb0e57f271585b8ea9bb355c7c79020716cfb9b1183ef3ab20e37d57a6b9d7477609aee6e122a4cf51427325250c7d0e509289444c9b3a648f1d71035d2ed65b0e3cdd0cbae8bf2d0b227812cbb360987255cc744110c453baa4fcd610928d809810e4b7ed1a8fd991f06aa6248204797e36a6a73b70a2559c09ead686945ba246ab66e5edd8044b4c6de3fcf2a89441ac66272fd8fb330ef8190579b3684596c960bd596eea520a56a8d650f563aad27409960dca63d3e688611ea5e22f4415cf9538d51a200c27034272968a264ed6540c84838d89f72c24461aad6d26f59ecaba9acbbb317b66d902f4f292a36ac1b639c637ce343117b659622245317b49eeda0c6258f100d7d961ffb138647e92ea330faeea6dfa31c7a84dc3bd7e1b7a6c7178af36879018e3f252107f243d243dc7339d5684c8b0378bf30244da8c87c843f5e56eb4c5e8280a2b48052cf93b16499a66db7cca71e4599426f7d461e66f99882bd89fc50800becca62d6c74116dbd2972fda1fa80f85df881edbe5a37668936b335583b599186dc5c6918a396fa48a181d6b6fa4f9d62d513afbb992f2b992f67f8afe67f76913fa388cb5630c8ca01e0c65d11c66a1e2ac4c85977b7c7a6999bbf10dc35ae69f5515614636c0b9b68c19ed2e31c0b3b66763038ebba42f3b38edc0399f3a9f23faa63978c317fc9fa66a73f60f0504de93b5b845e275592c12335ee340bbc4fddd502784016e4b3be7ef04dda49f4b440a30cb5d2af939828fd4ae3794e44f94df5a631ede42c1719bfda";
    const tag_hex = "bf0253fe5175be898e750edc53370d2b";
    const ad_hex = "17030302a2";
    const nonce_hex = "5d313eb2671276ee13000b30";
    const key_hex = "3fce516009c21727d0f2e4e86ee403bc";

    var c: [c_hex.len/2]u8 = undefined;
    var tag: [tag_hex.len/2]u8 = undefined;
    var ad: [ad_hex.len/2]u8 = undefined;
    var nonce: [nonce_hex.len/2]u8 = undefined;
    var key: [key_hex.len/2]u8 = undefined;

    _ = try std.fmt.hexToBytes(&c, c_hex);
    _ = try std.fmt.hexToBytes(&tag, tag_hex);
    _ = try std.fmt.hexToBytes(&ad, ad_hex);
    _ = try std.fmt.hexToBytes(&nonce, nonce_hex);
    _ = try std.fmt.hexToBytes(&key, key_hex);

    var m: [c.len]u8 = undefined;
    try std.crypto.aead.aes_gcm.Aes128Gcm.decrypt(&m, &c, tag, &ad, nonce, key);
}
```
```console
Test [1/1] test.Aes128Gcm... FAIL (AuthenticationFailed)
/home/naoki/.local/zig-master/lib/std/crypto/aes_gcm.zig:95:17: 0x2146c3 in decrypt (test)
                return error.AuthenticationFailed;
                ^
./ghash.zig:23:5: 0x213e5b in test.Aes128Gcm (test)
    try std.crypto.aead.aes_gcm.Aes128Gcm.decrypt(&m, &c, tag, &ad, nonce, key);
    ^
0 passed; 0 skipped; 1 failed.
error: the following test command failed with exit code 1
```

### Expected Behavior

Succesfully decrypted.
```console
All 1 tests passed
```

**Comment by jedisct1** [comment]
Thank you for reporting this!

**Comment by naoki9911** [comment]
Thanks! But below case seems not passing test

```zig
const std = @import("std");

test "Aes128Gcm" {
    const c_hex = "a23f7054b62c94d0affafe8228ba55cbefacea42f914aa66bcab3f2b9819a8a5b46b395bd54a9a20441e2b62974e1f5a6292a2";
    const tag_hex = "977014bd1e3deae63aeebb21694915e4";
    const ad_hex = "1703030043";
    const nonce_hex = "5b78923dee08579033e523d9";
    const key_hex = "17422dda596ed5d9acd890e3c63f5051";

    var c: [c_hex.len/2]u8 = undefined;
    var tag: [tag_hex.len/2]u8 = undefined;
    var ad: [ad_hex.len/2]u8 = undefined;
    var nonce: [nonce_hex.len/2]u8 = undefined;
    var key: [key_hex.len/2]u8 = undefined;

    _ = try std.fmt.hexToBytes(&c, c_hex);
    _ = try std.fmt.hexToBytes(&tag, tag_hex);
    _ = try std.fmt.hexToBytes(&ad, ad_hex);
    _ = try std.fmt.hexToBytes(&nonce, nonce_hex);
    _ = try std.fmt.hexToBytes(&key, key_hex);

    var m: [c.len]u8 = undefined;
    try std.crypto.aead.aes_gcm.Aes128Gcm.decrypt(&m, &c, tag, &ad, nonce, key);
}
```

**Comment by naoki9911** [comment]
And I've noticed that [this line](https://github.com/ziglang/zig/blob/2897641fb95a614ae61a444e43a0555804701910/lib/std/crypto/aes_gcm.zig#L38) may have to be
```zig
const block_count = (math.divCeil(usize, ad.len, Ghash.block_length) catch unreachable) + (math.divCeil(usize, c.len, Ghash.block_length) catch unreachable) + 1;
```
This enables some plaintexts to be encrypted correctly.

**Comment by jedisct1** [comment]
@naoki9911 Right, we need to provision for the block with the sizes as well.

**Comment by naoki9911** [comment]
Thanks!
Above decryption error was fixed with [this PR(#13527)](https://github.com/ziglang/zig/pull/13527)
